src/main/scala/is/hail/expr/ir/ExtractAggregators.scala

```diff
@@ -0,0 +1,174 @@
+package is.hail.expr.ir
+
+import is.hail.asm4s._
+import is.hail.annotations._
+import is.hail.annotations.aggregators._
+import is.hail.expr.{Type, RegionValueAggregator, TAggregable, TBoolean, TInt32, TInt64, TFloat64, TFloat32, TArray}
+import is.hail.utils._
+
+import scala.language.existentials
+import scala.language.postfixOps
+
+object ExtractAggregators {
+
+  private case class IRAgg(in: In, typ: Type, agg: RegionValueAggregator) { }
+
+  def apply(ir: IR, tAggIn: TAggregable): (IR, RegionValueAggregator) = {
+    val (ir2, aggs) = extract(ir, tAggIn)
+    val combined = new ZippedRegionValueAggregator(aggs.map(_.typ), aggs.map(_.agg))
+    // mutate the type of the input IR node now that we know what the combined
+    // struct's type is
+    aggs.foreach(_.in.typ = combined.typ)
+    (ir2, combined)
+  }
+
+  private def extract(ir: IR, tAggIn: TAggregable): (IR, Array[IRAgg]) = {
+    val ab = new ArrayBuilder[IRAgg]()
+    val ir2 = extract(ir, ab, tAggIn)
+    (ir2, ab.result())
+  }
+
+  private def extract(ir: IR, ab: ArrayBuilder[IRAgg], tAggIn: TAggregable): IR = {
+    def extract(ir: IR): IR = this.extract(ir, ab, tAggIn)
+    ir match {
+      case Ref(name, typ) =>
+        assert(typ.isRealizable)
+        ir
+      case _: AggIn | _: AggMap | _: AggFilter | _: AggFlatMap =>
+        throw new RuntimeException(s"Aggregable manipulations must appear inside the lexical scope of an Aggregation: $ir")
+      case AggSum(a, typ) =>
+        val tChildAgg = a.typ.asInstanceOf[TAggregable]
+        val tElement = tChildAgg.elementType
+        val in = In(0, null)
+
+        val fb = FunctionBuilder.functionBuilder[MemoryBuffer, Long, Boolean, RegionValueAggregator, Unit]
+        val aggregator = fb.getArg[RegionValueAggregator](4)
+        val (seqOp, agg) = tElement match {
+          case _: TBoolean => ((v: Code[_], mv: Code[Boolean]) =>
+            mv.mux(
+              Code.checkcast[RegionValueSumBooleanAggregator](aggregator).invoke[Boolean, Boolean, Unit](
+                "seqOp", coerce[Boolean](defaultValue(tElement)), true),
+              Code.checkcast[RegionValueSumBooleanAggregator](aggregator).invoke[Boolean, Boolean, Unit](
+                "seqOp", coerce[Boolean](v), false)),
+            new RegionValueSumBooleanAggregator())
+          case _: TInt32 => ((v: Code[_], mv: Code[Boolean]) =>
+            mv.mux(
+              Code.checkcast[RegionValueSumIntAggregator](aggregator).invoke[Int, Boolean, Unit](
```

Comment:
Kind of unhappy about all this duplication. I think to avoid it I have to work one level deeper in Code by explicitly passing `TypeInfo` to the `Invokeable` stuff.

Reply:
It is unnecessary.  You can just cast to `RegionValueAggregator` in all cases since  you're calling a function from its interface.  Just the constructor sum aggregator object needs to vary.

Reply:
I must misunderstand something because I'm pretty sure I'm _not_ calling a function from `RegionValueAggregator`'s interface. For example, here I call the `RegionValueSumIntAggregator`'s `seqOp(int, boolean)` method. The `RegionValueAggregator` interface takes an offset to a value. I didn't want to unnecessarily load an intermediate value into the region just so I can get an offset to hand to the aggregator.

Reply:
Sorry, I thought you were calling the `seqOp(MemoryBuffer, Long, Boolean)` variant.  Is there any reason you can't do that?

Also, the mux here seems unnecessary.  Isn't `v` some dummy value already when `mv` is false?

Reply:
I can, but an intermediate value (any non-compound thing) will have to be written into the region so as to get an offset that I can pass to that interface. It seems silly to, for example, load the result of simple arithmetic (performed on the JVM stack) into the region and then unload it on the other side of the RegionValueAggregator interface.

Executing `v` when `mv` is `true` seems like a bad idea. What if `v` is `ArrayLength(NA(TArray(TInt32())))`. If the region is empty, then I'll try to loadLength from an array at offset 0, but `loadInt` requires the region is at least 4 bytes large. I think the crux of the issue is that we don't get undefined behavior when we read bad memory addresses, we get assertion errors. An even more problematic example is `ArrayRef(NA(TArray(TInt32())), I32(0)`. The offset to the missing array will get dummy value 0, and `loadElement` will happily try to read the bytes at 0 as the length of the array which of course could lead to read attempts outside the region.

Reply:
You're right, my bad.  I was thinking the thing had an offset already and you should hold on to it, but that's clearly wrong because it might be the intermediate result of an expression.  This seems like the standard type specialization duplication.  Maybe you could do some type-driven seqOp lookup to avoid the duplication, but honestly, this looks fine to me now.

This comment on Compile makes the second point clear:

 > gaurd the the evaluation of value by missingness

It does make me wonder if this is the right choice.  In general, you had an if in the precompute Code to test the missingness of the inputs, and now you force yourself to have a second one in the consumption.  I'm OK with coming back to this when we work on optimized the generated code.

Reply:
The problem arises with values that _must_ do some computation to know if they are missing. For example, an array reference must know the value of the array to know if it is itself missing. I can either:

 - duplicate the evaluation of (and the bytecode for) the array

 - duplicate the missingness check

---

I think this situation would be resolved if the compiler was "CPS'ed" (using two "continuations"). The ArrayRef implementation would lay down the byte code to evaluate the array, then mux on the element's missingness:
 -  in the missing case it calls the "My value is missing" "continuation" (whose type is `() => Code[_]`)
 - if the element was non-missing, call the "here's my value" continuation (whose type is `(Code[_]) => Code[_]`) with the byte code to load the element

This mirrors how one would handle exceptions in a simple CPS'ed Scheme compiler, which is kinda how missingness works anyway.

Reply:
I agree with your second comment that compiling with dual continuations is likely to generate better code.

I don't fully understand the first case.  You can just have a default initialized value for the array element value.  For example, imagine you're compiling `gs.map(g => a[g.DP]).sum()`, then you can do:

```
  var e, me // the element of a, default-initialized
  // evaluate a, ma
  if (ma)
    me = false
  else
    e = a(a.DP)
    me = true
  aggs(0).seqOp(e, me)  // don't need a second test because e is ignored if me = true
```

Reply:
This (two continuation compilation) is essentially what I tried to do for Booleans in `Code`.  It's much better because you don't actually need branches on the intermediate missing bits and will almost certainly generate better code.  It dominates the usual approach because you can always set a missing bit on the two branches and jump to common code (like if you need to pass a value out or store it.)

Reply:
The byte code example you give works if I always pass values via local variables. In such a situation I can always set the value to a default value in the setup code that runs first.

One of the design goals I followed was that the compiler only stores a value in a local variable if it a) is the bound value in a let or b) is _definitely_ used more than once. For example, I don't store the value of `ArrayRef` in a local variable, (it's just a `region.loadType(tarray.loadElement(...))`), so I'd have to duplicate the missingness check in the code for the value to return a default value.

Storing all returned values in local variables seems a little silly, especially in large compositions of arithmetic expressions where every binary operation will get stored and then loaded.

The only place this is an issue is at method call boundaries like this (which never occur in regular IR) because method arguments are evaluated eagerly.